### PR TITLE
FixCommand - Support rules with params

### LINKED ISSUE #2423: [Question] How to execute rules with params via command line
I am trying to apply only a rule, without create a .php_cs config file, for example:

    $ php vendor/friendsofphp/php-cs-fixer/php-cs-fixer --rules=array_syntax mysrc

But that will take the default value for array_syntax value (which is syntax => long).

I would like use params via command line, something like:

    $ php vendor/friendsofphp/php-cs-fixer/php-cs-fixer fix --rules=array_syntax[syntax=>short] mysrc

That should convert all the arrays with syntax short (that is array() to [] ) which is more syntax sugar and clean for improve readability. But this obviously doesn't work and it doesn't take the "syntax" param as supported.

Alternative ways (using quoted parameters) could be this one:

 $ php vendor/friendsofphp/php-cs-fixer/php-cs-fixer fix --rules=array_syntax['syntax' => 'short'] mysrc

In fact this produces a weird exception (probably a bug) assuming that is a path:

    [PhpCsFixer\ConfigurationException\InvalidConfigurationException]  
  The path "=" is not readable.

But this is mostly a regresion, because in previous versions the documentation says that you can use short_array_syntax that was "deprecated" as you can read in [UPGRADE](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/UPGRADE.md) file, but it is only possible to use the values via config file and don't via parameters in command line?

### PR DESCRIPTION
**Pull request by ptcong**
According to issue #2423, i made this PR.
```
--rules='{"@PSR2": true, "class_keyword_remove": true, "concat_space": {"spacing": "none"}, "dir_constant": true}'
```

Closes #2423, #2502, #2510

### CODE REVIEW

**Comment by SpacePossum** [comment]
Hi and thanks for the PR!
Please provide a utest for some cases. For example I think if someone passes a string that is not valid json this will result in an exception/error being raised.

**Comment by ptcong** [comment]
When json_decode a string, it will returns null if failed, doesn't throw anything. So if user passes a string, it will run the code block under.

**Comment by keradus** [comment]
Hi. Still, utest is required

**Comment by ptcong** [comment]
I've just added test for this case.

**Comment by SpacePossum** [comment]
(Please fix the tests by fixing the code style of the project itself.)

I think it would be nice improvement to allow JSON (that decodes to `array`), or rules comma separated list of fixer names.
This PR aims for that, however it would be nice to check if either one of two formats is used.
Now if you pass invalid JSON (or valid but resolves to non-array) the user does not get an error message that makes sense, because the tool will try to explode the invalid JSON input. WDYT?

File: src/Console/ConfigurationResolver.php [hunk 1]
```diff
@@ -588,6 +588,12 @@ private function parseRules()
             return $this->getConfig()->getRules();
         }
 
+        $jsonRules = json_decode($this->options['rules'], true);
+
+        if (is_array($jsonRules)) {
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: exit on invalid user input instead of silently going forward

Comment (ptcong) [inline src/Console/ConfigurationResolver.php]: Ok.

File: src/Console/ConfigurationResolver.php [hunk 2]
```diff
@@ -588,10 +588,15 @@ private function parseRules()
             return $this->getConfig()->getRules();
         }
 
-        $jsonRules = json_decode($this->options['rules'], true);
+        $rules = trim($this->options['rules']);
 
-        if (is_array($jsonRules)) {
-            return $jsonRules;
+        if (in_array($rules[0], array('[', '{', '"', "'"), true)) {
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: valid json object must starts from `{`

File: src/Console/ConfigurationResolver.php [hunk 3]
```diff
@@ -588,10 +588,15 @@ private function parseRules()
...
+            $rules = json_decode($rules, true);
+            if (!is_array($rules)) {
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: if it was decoded we already know it's an assoc object as it has to starts with `{` (due to prev comment).
then, consume `json_last_error` instead

File: src/Console/ConfigurationResolver.php [hunk 4]
```diff
@@ -588,10 +588,15 @@ private function parseRules()
...
+                throw new InvalidConfigurationException('Invalid rules input.');
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: details to the user are needed

File: src/Console/ConfigurationResolver.php [hunk 5]
```diff
@@ -590,10 +590,11 @@ private function parseRules()
 
         $rules = trim($this->options['rules']);
 
-        if (in_array($rules[0], array('[', '{', '"', "'"), true)) {
+        if ($rules[0] === '{') {
             $rules = json_decode($rules, true);
-            if (!is_array($rules)) {
-                throw new InvalidConfigurationException('Invalid rules input.');
+            if ($error = json_last_error()) {
+                $error = function_exists('json_last_error_msg') ? json_last_error_msg() : "Error $error";
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: drop checking method existence please

File: src/Console/ConfigurationResolver.php [hunk 6]
```diff
@@ -588,6 +588,17 @@ private function parseRules()
             return $this->getConfig()->getRules();
         }
 
+        $rules = trim($this->options['rules']);
+
+        if ($rules[0] === '{') {
+            $rules = json_decode($rules, true);
+            if ($error = json_last_error()) {
```
Comment (SpacePossum) [inline src/Console/ConfigurationResolver.php]: Although I like it, please don't assign inside the condition.
And please use `sprintf` to construct the exception message.

File: tests/ConfigTest.php [hunk 7]
```diff
@@ -25,6 +26,41 @@
  */
 final class ConfigTest extends \PHPUnit_Framework_TestCase
 {
+    public function testConfigRulesUsingJson()
```
Comment (SpacePossum) [inline tests/ConfigTest.php]: nice cases :)
when you are done with the creating all tests please split up the cases in separate methods, pref. using dataproviders where handy.

File: src/Console/ConfigurationResolver.php [hunk 8]
```diff
@@ -588,6 +588,17 @@ private function parseRules()
...
+                throw new InvalidConfigurationException("Invalid JSON rules input: Error $error");
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: drop checking method existence please
add `symfony/polyfill-php55` to requirements

File: src/Console/ConfigurationResolver.php (hunk 6 above)
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: `if (null !== json_last_error())`

File: src/Console/ConfigurationResolver.php (hunk 8 above)
Comment (ptcong) [inline src/Console/ConfigurationResolver.php]: But the library still supports PHP 5.3 while `json_last_error_msg` required 5.5

> PHP needs to be a minimum version of PHP 5.3.6.

Comment (keradus) [inline src/Console/ConfigurationResolver.php]: so, for the 3rd time,

> add `symfony/polyfill-php55` to requirements

after this, `json_last_error_msg` will work on 5.3

Comment (SpacePossum) [inline src/Console/ConfigurationResolver.php]: >But the library still supports PHP 5.3 while json_last_error_msg required 5.5

yes, thats why `add symfony/polyfill-php55 to requirements`  ;)

Comment (ptcong) [inline src/Console/ConfigurationResolver.php]: Thanks

File: composer.json [hunk 9]
```diff
@@ -23,7 +23,8 @@
         "symfony/finder": "^2.2 || ^3.0",
         "symfony/polyfill-php54": "^1.0",
         "symfony/process": "^2.3 || ^3.0",
-        "symfony/stopwatch": "^2.5 || ^3.0"
+        "symfony/stopwatch": "^2.5 || ^3.0",
+        "symfony/polyfill-php55": "^1.3"
```
Comment (keradus) [inline composer.json]: how did you add it?
why https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/composer.json#L37 didn' t work ?

Comment (ptcong) [inline composer.json]: Sorry, i just added it manually :D

File: tests/ConfigTest.php [hunk 10]
```diff
@@ -25,6 +26,47 @@
...
+        $config = new Config();
+        $configResolver = new ConfigurationResolver(
+            $config, array(
+                'rules' => '{"array_syntax": {"syntax": "short"}, "cast_spaces": true}',
+            ),
+            getcwd()
+        );
+        $fixers = $configResolver->getFixers();
+        $this->assertInstanceOf(\PhpCsFixer\Fixer\ArrayNotation\ArraySyntaxFixer::class, $fixers[0]);
+        $this->assertInstanceOf(\PhpCsFixer\Fixer\CastNotation\CastSpacesFixer::class, $fixers[1]);
```
Comment (keradus) [inline tests/ConfigTest.php]: * please do not rely on the order here, `getFixers` may return fixers in different order depends on priority of the fixers
* please check if the config was applied

File: tests/ConfigTest.php [hunk 11]
```diff
@@ -25,6 +26,47 @@
...
+    public function testConfigRulesUsingInvalidJson()
+    {
+        $config = new Config();
+        $configResolver = new ConfigurationResolver(
+            $config, array(
+                'rules' => '{blah',
+            ),
+            getcwd()
+        );
+        $this->expectException(\PhpCsFixer\ConfigurationException\InvalidConfigurationException::class);
```
Comment (keradus) [inline tests/ConfigTest.php]: please set up the expectation on top of test method

**Comment by keradus** [comment]
:+1; after travis will be fixed (one cannot use `::class` on lower PHP versions)

**Comment by SpacePossum** [comment]
:+1:
(after test is green)

**Comment by keradus** [comment]
Thank you @ptcong.

**Pull request closed without merging**